File: kotlin/roman-numerals/src/main/kotlin/RomanNumerals.kt
object RomanNumerals {
    private val decimalToRoman = mapOf(
            1000 to "M", 900 to "CM", 500 to "D", 400 to "CD",
             100 to "C",  90 to "XC",  50 to "L",  40 to "XL",
              10 to "X",   9 to "IX",   5 to "V",   4 to "IV",
               1 to "I"
    )

    fun value(decimal: Int): String {
        var n = decimal
        return decimalToRoman
                .entries
                .fold(StringBuilder()) { roman, entry ->
                    while (n >= entry.key) {
                        roman.append(entry.value)
                        n -= entry.key
                    }
                    roman
                }
                .toString()
    }
}

File: elixir/roman-numerals/lib/roman_numerals.ex
defmodule RomanNumerals do
  @map [
    M: 1000, CM: 900, D: 500, CD: 400,
    C: 100,  XC: 90,  L: 50,  XL: 40,
    X: 10,   IX: 9,   V: 5,   IV: 4,
    I: 1
  ]

  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(n), do: do_roman(@map, n, "")

  defp do_roman(_, 0, digits), do: digits

  defp do_roman([{r, d} | map], n, digits) when n >= d do
    do_roman([{r, d} | map], n - d, digits <> to_string(r))
  end

  defp do_roman([_ | map], n, digits) do
    do_roman(map, n, digits)
  end
end
